Improved the Red Alert dog behavior

Mailaender opened the pull request:
Fixes https://github.com/OpenRA/OpenRA/issues/7506. I redid this from scratch and tried to match RA95 behavior taking https://youtu.be/4ZyAlXOY0aM?t=2m for reference still avoiding legacy limitations.

Instead of leaping onto a targetable actor and killing everything in that cell, I separate leap movement and attack execution. This means the path finder now makes sure that an attack can be executed so jumping over cliffs and other glitches are gone.

The insta-kill is also not a hard-coded hack, but using the regular weapon system. This means we can also armor soldiers accordingly so they only take damage from a hit for a more fine grained balancing. For now the weapon is a high speed bullet like all the other melee attacks, but we can change this into a close combat "projectile" later on to avoid the overhead.

I also tried to adhere to our modular https://github.com/OpenRA/OpenRA/wiki/Actor-Upgrades system instead of putting everything into a monolithic one-purpose trait. The `UpgradeOnTarget` trait can be used to recreate http://cnc.wikia.com/wiki/Terrorist_(Generals) style units and I guess creative modders will find a lot of uses for `UpgradeOnAttack` I didn't even think of yet.

Mailaender committed:
Add UpgradeOnTarget.

Mailaender committed:
Add UpgradeOnAttack.

Mailaender committed:
Rewrite Leap and AttackLeap

so that Leap only does the movement
and AttackLeap executes it using the weapon system.

pchote commented on OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnAttack.cs:
```diff
@@ -0,0 +1,56 @@
+		void INotifyAttack.PreparingAttack(Actor self, Target target, Armament a, Barrel barrel)
+		{
+			foreach (var up in info.PreparingAttackUpgrades)
+				manager.GrantUpgrade(self, up, this);
```
Is it possible to call PreparingAttack without a corresponding Attacking? Need to avoid mismatched calls.

pchote replied:
This isn't possible, except when the actor is killed in the middle.  But the actor is then dead, so we don't care.

Mailaender commented on OpenRA.Mods.RA/Traits/Attack/AttackLeap.cs:
```diff
@@ -9,49 +9,44 @@
 {
-	[Desc("Dogs use this attack model.")]
-	class AttackLeapInfo : AttackFrontalInfo
+	[Desc("Move onto the attack then execute the attack.")]
```
Typo: `target`

pchote commented on mods/ra/weapons/other.yaml:
```diff
@@ -185,9 +185,12 @@ TTankZap:
+	Range: 1c512
+	Projectile: Bullet
+		Speed: 1c682
 	Warhead@1Dam: SpreadDamage
```
Dogs have small mouths, so spread isn't a great model here.  Someone needs to bite the bullet and add a proper direct-damage warhead (bypassing spread and positions altogether), would you like to volunteer?

WolfGaming replied:
I can't believe we still don't have a direct-damage warhead...

GraionDilach replied:
The problem is that only the `*Zap`s and `Missile` tracks the target, (and `Missile` will never hit it due to CloseEnough). Even with a direct-damage warhead present, it will be still useless because all projectiles will miss it - some doesn't even use `GuidedTarget` to begin with!

Mailaender replied:
We now do have it. https://github.com/OpenRA/OpenRA/pull/11715/commits/979b328e6e931c22d06480ea3e73b887cd6ef231

Mailaender commented:
Small update with feedback from yesterdays IRC chat:
* It should now not crash when the dogs are not in the world.
* The dogs are immobile during the eating step.

MustaphaTR commented:
> It should now not crash when the dogs are not in the world.

No, it still does.

Also 1c512 as range is too short. They can't always attack from some ways. In original game they have 2.2 range.

Mailaender commented:
Can you post a stacktrace and a way to reproduce the crash?

MustaphaTR commented:
I killed an enemy dog with my dog.

Mailaender committed:
Add an instant hit direct on target "projectile".

Mailaender committed:
Add an actor damage warhead that only counts direct hits.

Mailaender committed:
Change dog behavior to match RA95 more closely:

Speedup when an actor is targeted.
Become immobile for a while shortly after an attack.
Setup unused attack sequences.

Mailaender commented:
I updated this again:
* Fixed the crash when 2 dogs attacked each other simultaneously.
* Added a direct-damage warhead and a melee projectile.

GraionDilach commented:
It would be cool if `UpgradeOnAttack` would have some sort of `Armament` filtering, say with a `Name` field, but I can understand if that gets dismissed.

Mailaender commented:
Yes, that should be easy. Also a target type filter for `UpgradeOnTarget` would be possible, but I don't have a test case in this pull request as dogs only have one suitable target type and only one armament so this is out of scope and the pull request already touches a lot of things and will probably have a very slow review process. I suggest you create followup pull request based on this on your own.

MustaphaTR commented on mods/ra/weapons/other.yaml:
```diff
@@ -186,10 +186,10 @@ TTankZap:
 	ValidTargets: Infantry
 	ReloadDelay: 10
-	Range: 3c0
+	Range: 2c256
```
That's `2.25` not `2.2` but i'm not sure if it really matters or not.

Mailaender replied:
I don't really know how the original `RULES.INI` values have to be converted to match the OpenRA distances. Based it on a rough visual comparison. Also we probably want to rebalance this properly. I never saw someone on @cncnet actually train dogs.

GraionDilach commented:
Alright, I have no changes requested. Code looks good otherwise, but I couldn't test this and I'm gone for camping for a week from tomorrow onwards, so :+0.5: from here.

Biofreak1987 commented:
the jumpattack and eating animations still seem to have some problems:
- jumpattack animations never play
- eating animations only play "sometimes" when attacking other dogs

pchote commented:
This doesn't work at all for me... the dogs will run up next to my infantry and then stop.

pchote commented on OpenRA.Mods.RA/Traits/Attack/AttackLeap.cs:
```diff
@@ -9,49 +9,47 @@
-				return;
-
-			if (!target.IsInRange(self.CenterPosition, a.MaxRange()))
+			if (IsAttacking || !mobile.CanMoveFreelyInto(self.World.Map.CellContaining(target.CenterPosition)))
```
This should be moving to the target infantry's sub-cell rather than the center of the cell.

pchote commented on OpenRA.Mods.RA/Activities/Leap.cs:
```diff
@@ -9,70 +9,53 @@
-				self.World.ActorMap.GetActorsAt(mobile.ToCell, mobile.ToSubCell)
-					.Except(new[] { self }).Where(t => weapon.IsValidAgainst(t, self))
-					.Do(t => t.Kill(self));
+				mobile.SetPosition(self, pos);
```
This ignores the sub-cell positioning, which I expect will look wrong.

Mailaender commented:
I updated this with sub-cell positioning in mind although I thought `WPos` already takes care about it even more precisely. Will remove https://github.com/OpenRA/OpenRA/pull/11715/commits/979b328e6e931c22d06480ea3e73b887cd6ef231 as scope creep if problems persists.

GraionDilach commented on OpenRA.Mods.Common/Projectiles/MeleeAttack.cs:
```diff
@@ -0,0 +1,50 @@
+namespace OpenRA.Mods.Common.Projectiles
+{
+	[Desc("Simple invisible direct on target actor fake projectile.")]
+	public class MeleeAttackInfo : IProjectileInfo
```
I'd rename this to InstantHitTarget or something along the line because that's what it is and can be used for other things besides melee.

Phrohdoh replied:
👍 Due to melee implying a very small distance.

Mailaender commented:
Gave the projectile a more generic name as suggested.

I also added a new leap upgrade and made the attack upgrade delay configurable at trait level. This makes balancing more independent from how the artwork works and resolves all regressions introduced during review compared to the initial submission.

abcdefg30 commented:
Needs a rebase.

Mailaender commented:
Rebased.

pchote commented:
Infantry are killed long-before the dog reaches them: https://www.dropbox.com/s/m96wjtb6isznhb2/dog.mov?dl=0

It really puts me off reviewing PRs when, over multiple PR iterations, I find merge-blocking bugs within seconds of testing them.

Mailaender commented:
That video gives me error 500. It is a bit sad that I hadn't saved the initial revision of the pull request before the review requests for more features and complexity began.

GraionDilach commented:
The traits seem fine at first glance tbh, so I could try giving you a helping hand at the yaml setup, if you'd like to - prolly only on Wednesday though.

Mailaender commented:
Yes, that would be greatly appreciated. I probably don't have an eye for the details and the timing.

pchote commented:
I guess we don't want to match the RA95 behaviour *too* well:

https://www.reddit.com/r/IAmA/comments/19epje/i_am_joe_bostic_22_year_game_developer_11_years/c8o39zv

GraionDilach commented:
Haven't found time, but I guess the blinking frame comes in due to no upgrade is applied during PreparingAttack - granting rip should fix the issue I guess.

reaperrr commented on OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnTarget.cs:
```diff
@@ -0,0 +1,61 @@
+			manager = self.Trait<UpgradeManager>();
+		}
+
+		public void Tick(Actor self)
```
`void ITick.Tick`, please.

reaperrr commented on OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnAttack.cs:
```diff
@@ -0,0 +1,59 @@
+		[UpgradeGrantedReference] public readonly string[] AttackingUpgrades = { };
+
+		[Desc("How long the timed attacking upgrade should last.")]
+		public readonly int AttackingDelay = 0;
```
I think in this case `AttackingDuration` would be better. `Delay` is misleading, in my opinion.

reaperrr commented:
Needs rebase.

I'll try to review this sometime next weekend.

pchote commented:
There doesn't appear to be any interest in driving this forward, so moving to the future milestone and closing.

pchote commented:
This has been superseded by #13513, so removing from the milestone.
